feat: New Notification: Slow Response Time w/ Resend Interval

stephenpapierski opened the pull request:
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

This new feature allows notifications to be configured surrounding the response time of a given service.

Resolves https://github.com/louislam/uptime-kuma/issues/1813
Closes https://github.com/louislam/uptime-kuma/pull/1878
Fixes https://github.com/louislam/uptime-kuma/issues/4035

- Add `Resend Interval`
- Add `Last` Response Time Calculation Method
- Add `Threshold Calculation` with `Static Threshold` and `Relative to Avg. Response` Methods
- Also send notification when service returns to normal response time
- Add threshold line to ping chart
- Add `Slow` badge to monitor dashboard when service is responding slow (no badge when `Nominal`)
- Add `Slow/Nominal` Entries to events log
- Add `Slow/Nominal` toast messages
- Add number of `Slow` services to Dashboard Quick Stats
- Update discord and slack notification providers to provide `Slow/Nominal` notifications that looks similar to the existing `Up/Down` notifications. By default, the rest of the notification providers fall back to plain text notifications.

### Notification Control
| Control | Behavior |
| ------------- | ------------- |
| `Slow Response Notification`  | Enable slow response notifications |
| `Resend Notification if Slow Response X times consecutively` | Resend notification every x times the service responds slowly. A setting of 0 make it so that notifications are only sent when a service starts responding slowly or returns to a normal response time (situations 1 and 2 below) |

Notifications are sent during 3 events.
1. Response time is greater than specified threshold (Service started responding slowly)
2. Response time returns to less than specified threshold (Service returned to normal response time)
3. Response time remains slow **x** consecutive times, if `Resend Notification` is greater than 0 (Service still responding slowly)

### Response Time Calculation
There are multiple options for calculating the response time.
| Method | Additional Parameters | Description |
| ------------- | ------------- |------------- |
| Average  | `Window Duration` | Average response time during `Window Duration` |
| Max  | `Window Duration` | Maximum response time during `Window Duration` |
| Last | N/A | Response time of the last heartbeat |

### Threshold Calculation
There are also multiple options for calculating the threshold.
| Method | Additional Parameters | Description |
| ------------- | ------------- |------------- |
| Static Threshold | `Threshold` | Define a static threshold in milliseconds |
| Relative to Avg. Response | `Threshold Multiplier` | Calculate the threshold (`Threshold Multiplier` * 24-hour average response time). The goal here is that you don't have to tune your thresholds, it will simply tell you if your service is responding slower than "normal".

## Type of change

Please delete any options that are not relevant.
- Bug fix (non-breaking change which fixes an issue)
- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

### Feature Disabled (default)
"Slow Response Notification" checkbox below rest of "General" fields.
<img width="469" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/523dfd60-6a80-4da8-b996-91d9e30e8ab9">

### Feature Enabled (with default settings)
<img width="463" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/9a61f784-7f80-4354-ae00-88b7b2339d6a">

### Response Time Calculation

#### Average
<img width="469" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/8a0c28b8-f8d2-4619-af09-04c2392c6815">

#### Max
<img width="476" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/134e8703-b5ce-4710-a3e7-92eef948b836">

#### Last (Window Duration Hidden)
<img width="477" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/60143f72-5a60-4035-ab26-838eed5a889f">

### Threshold Calculation

#### Relative to Avg. Response
<img width="472" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/de4171f1-b30c-4d80-9c86-9d25842475d2">

#### Static Threshold
<img width="468" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/656c2281-5022-4c26-9eda-3d6c81ddfd0d">

### Resend Interval
<img width="465" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/0accc470-9aaa-4086-82ee-f20cca3ce131">

### Dashboard Changes
 - Add threshold line to ping chart
<img width="934" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/8dd4e2fa-7367-4aa4-a8e9-e33fc708ec50">

 - Add `Slow` badge to monitor dashboard when service is responding slow (no badge when `Nominal`)
<img width="937" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/1bc84526-069d-41bd-b194-26e8e3e32268">

 - Add `Slow`/`Nominal` entries to events log
![image](https://github.com/louislam/uptime-kuma/assets/5900005/6999d547-fdc0-4186-add3-44348ac928c8)

 - Add `Slow`/`Nominal` toast messages
![image](https://github.com/louislam/uptime-kuma/assets/5900005/303a2285-ce05-4bee-8ef4-8713b04e5c22)

 - Add number of `Slow` services to Dashboard Quick Stats
 <img width="947" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/a3b1d19c-28d6-4b34-85e2-57f8d41a58db">

### Notification Providers
#### Discord
<img width="456" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/02320460-7101-410c-8cc6-3d6d5fca5ff0">

#### Slack
<img width="638" alt="image" src="https://github.com/louislam/uptime-kuma/assets/5900005/de9908b6-daa8-449f-a70e-ecf830cb866c">

stephenpapierski committed:
Reimplement functionality from
https://github.com/louislam/uptime-kuma/pull/1878 with modern
knex_migration

stephenpapierski committed:
Add slow response notification resend interval to db/gui

stephenpapierski committed:
Add slow response notification resend interval, also notify when response time returns to normal

stephenpapierski committed:
Merge branch 'louislam:master' into feature-slow-response-notification

stephenpapierski committed:
Prevent slowResponseNotificationRange from being less than one Heartbeat Interval

stephenpapierski committed:
Fix behavior of slowResponseNotificationRetryInterval

stephenpapierski committed:
Add calculation method to just use last heartbeat

stephenpapierski committed:
Round response time to nearest ms when using average

stephenpapierski committed:
Merge branch 'louislam:master' into feature-slow-response-notification

stephenpapierski committed:
Clean up notification logic and messages

stephenpapierski committed:
Merge branch 'louislam:master' into feature-slow-response-notification

stephenpapierski committed:
Add threshold calculations methods

stephenpapierski committed:
Add threshold calculations methods

stephenpapierski committed:
Merge branch 'feature-slow-response-notification' of github.com:stephenpapierski/uptime-kuma into feature-slow-response-notification

stephenpapierski committed:
Don't check for slow response unless monitor is UP

stephenpapierski committed:
Add db defaults so existing entries don't show up empty

stephenpapierski committed:
ESLint fixes

stephenpapierski committed:
Add threshold line to ping chart

stephenpapierski commented:
Am I supposed to leave this as a draft during discussion, or move it to `Ready to review`?

Zaid-maker commented:
> Am I supposed to leave this as a draft during discussion, or move it to `Ready to review`?

Yeah leave it on draft during discussion its totally up to you as well.

stephenpapierski commented:
Ok, just wanted to make sure I didn't need to move to `Ready for review` to signal that it is in fact ready for review.

stephenpapierski committed:
Fix threshold line autoscaling y-axis range

CommanderStorm commented:
> move to `Ready for review` to signal that it is in fact ready for review

I think it is kind of implicit that if something is still a draft the author thinks he*she still needs to do something
=> won't get merged
=> might not get reviewd depending on time constriants

If Louis reviews these is up to his shedule, but he might take hints from above ^^

stephenpapierski commented:
I would agree if the guidelines didn't stop where they do.
![image](https://github.com/louislam/uptime-kuma/assets/5900005/121786aa-605e-45b2-b1b5-71664cdbc222)

I'll go ahead and move it to `Ready for review` and submit another pull request with a proposed update to the guidelines. Maybe something like...

8. If you have questions or need some further guidance, click "Change to draft".
9. Discussion
10. When you're ready for review by the maintainers, click "Ready for review"

stephenpapierski committed:
Add `Slow` display to monitor dashboard

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification

stephenpapierski committed:
lint fix

stephenpapierski committed:
Add slow response events to details dashboard

stephenpapierski committed:
Add slow response events to dashboard home

stephenpapierski committed:
Fix: Clear Events clears slow response events too

stephenpapierski committed:
Fix for important & pingImportant both true

stephenpapierski committed:
Add Slow to Dashboard Quick Stats

stephenpapierski committed:
Fix slow response status messages auto show up now

stephenpapierski committed:
Fix "slow" quick stats calculation

stephenpapierski committed:
Add slow response toast messages

stephenpapierski committed:
Adjust notification messagess

stephenpapierski committed:
Fix class typo

stephenpapierski committed:
Merge remote-tracking branch 'upstream/master' into feature-slow-response-visual-improvements

stephenpapierski committed:
Fixing a comment

stephenpapierski committed:
lint fix

stephenpapierski committed:
lint fix

stephenpapierski committed:
Fix: sendHeartbeatList now sends correct format

stephenpapierski committed:
Simplify logic due to bug fix

stephenpapierski committed:
Log language clarification

stephenpapierski committed:
oops, wrong numbers

stephenpapierski commented:
Do we think the defaults are ok? I've gone back and forth on a couple items
- Calculating the response time by averaging the last 5 minutes
  - My thought is that this helps remove outliers that will show up in `max` or `last`
- Calculating the threshold by multiplying the 24-hour average by 5x
  - I like the `Relative to Avg Response` method because it works without much tuning
  - I'm not sure if 5x is the right default though. It's hard to test/figure out what a typical performance degradation looks like without some more data. Feedback from someone who has logs during a network event like that might be helpful.

stephenpapierski committed:
Add `Slow`/`Nominal` to en.json

stephenpapierski committed:
Add threshold to pingMsg

stephenpapierski commented:
I also just wanted to give a heads up that I've been working on some visual elements surrounding the slow response monitoring (https://github.com/stephenpapierski/uptime-kuma/tree/feature-slow-response-visual-improvements).

I didn't want to make the initial pull request too big, but I can definitely go ahead and push it to this branch if we'd prefer to do it all at once.

Some of those visual elements include:

- Display threshold line on ping chart
- `Slow` badge next to `Up`/`Down` badge when service is responding slowly
- `Slow`/`Nominal` events show up in the Events table
- `Slow`/`Nominal` toast messages
- Number of `Slow` services added to Quick Stats

stephenpapierski committed:
Merge remote-tracking branch 'upstream/master' into feature-slow-response-visual-improvements

stephenpapierski committed:
Merge branch 'louislam:master' into feature-slow-response-notification

stephenpapierski committed:
Remove commented out code

stephenpapierski committed:
Update lock file

stephenpapierski committed:
Add support for SLOW/NOMINAL notification cards

stephenpapierski committed:
Add discord support for SLOW/NOMINAL notification cards

stephenpapierski committed:
Add slack support for SLOW/NOMINAL notification cards

stephenpapierski committed:
Move notification msg generation to send function

This better matches the existing model.

stephenpapierski committed:
lint fix

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-visual-improvements

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification-cards

stephenpapierski committed:
Merge branch 'feature-slow-response-visual-improvements' into feature-slow-response-notification

stephenpapierski committed:
Merge branch 'feature-slow-response-notification-cards' into feature-slow-response-notification

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification

stephenpapierski committed:
Add chartjs-plugin-annotation to package-lock.json

Ayyaaya commented:
FEISHU no time

CommanderStorm commented:
@Ayyaaya What are you trying to convey?

blevkovych commented:
bump?

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification

stephenpapierski committed:
Combine and update naming of db migration

stephenpapierski committed:
Update address logic to use extractAddress

stephenpapierski committed:
Remove unused dayjs reference

stephenpapierski committed:
Fix badge margin

stephenpapierski committed:
Fix Notification.send call

stephenpapierski committed:
Fix color of Slow quick stats when there are 0 slow

stephenpapierski committed:
Fix vue warning for accessing heartbeatList

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

stephenpapierski committed:
Fix casing of alert_color to alertColor

stephenpapierski commented:
> I just found this. How can I test this feature to provide feedback? Looks great!

https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests

stephenpapierski committed:
Fix typo Legent -> Legend

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification

louislam commented on server/notification.js:
```diff
@@ -215,7 +215,20 @@ class Notification {
+                );
+                return this.providerList[notification.type].send(notification, msg);
+            } else {
+                return this.providerList[notification.type].send(notification, msg, monitorJSON, heartbeatJSON);
```
I think it would be better to introduce a new method, naybe `.sendSlowMessage(...)`

louislam reviewed (changes requested):
Just checked out the pull request and test it.

I think function wise, it is working as expected. I don't see any bugs yet.

In terms of UI/UX, this pr added 5 options, and I think they are a bit hard for me to understand honestly. Maybe we need to think a better presentation.

For me, as a normal user, I think I would just expect to check the checkbox, and input threshold (ms), and done.



The pull request is a bit large, may still need some time to test deeply.

stephenpapierski commented:
> In terms of UI/UX, this pr added 5 options, and I think they are a bit hard for me to understand honestly. Maybe we need to think a better presentation.
> 
> For me, as a normal user, I think I would just expect to check the checkbox, and input threshold (ms), and done.

I could get on board with simplified options. What you're describing would be a Response Time Calculation of `Last` and a Threshold Calculation of `Static Threshold`. Below are my justification for why I added the features. But I understand it might add too much complexity. if that's the decision, I can simplify it to just have a static threshold and remove the extra controls.

## Response Time Calculation
I added options other than `Last` since I wanted to avoid a single slow response triggering the notification. `Average` over a duration allows for a little more flexibility to avoid a single anomaly.

## Threshold Calculation
`Static` is the simple one where it's just a fixed value. `Relative to Average Response` came from the desire to avoid actually having to know how fast something responds. I just want to know if it's responding slower than normal.

stephenpapierski commented on server/model/monitor.js:
```diff
@@ -1613,6 +1640,238 @@ class Monitor extends BeanModel {
+                    AND monitor_id = ?
+                    AND status = ?
+                    `,
+                        [afterThisDate.toISOString(), monitor.id, UP]
```
This is only relevant if we're keeping the `average` or `max` Response Time Calculations around, so I'll hold off on that for now.

stephenpapierski committed:
Merge branch 'master' into feature-slow-response-notification

stephenpapierski committed:
Add comment differentiating SLOW/NOMINAL status values
